### DISCUSSION TITLE
Is there a good way to find out iif a command exists?

### DISCUSSION THREAD

**Discussion by btsimonh** [arendst/Tasmota#18704] [Q&A] [answered]
Hi, 
I'm looking at something which will execute a command from C, where the command may or may not have been added from Berry.
e.g. I'd like to call into Berry when motion on the webcam is detected.

So from the webcam code, I could execute a command, but I'd rather not execute it if it does not exist..... 

Is there a simple way in Tas C to determine if a command exists?

br,

Simon

**Comment by btsimonh** [comment]
Or should I be using the event command? - although this is in rules, which I'm not building because it conflicts with script :(.

**Reply by s-hadinger** [reply]
No there isn't.

**Reply by btsimonh** [reply]
what would you recommend to make something in Berry be called each time a frame was available in the webcam stuff, and/or each time a motion calculation has been done, apart from polling?

**Reply by sfromis** [reply, marked as answer]
I'd stop thinking of it as a "command", but instead have the Berry code set up a callback, similar to what is done with tasmota.add_rule, mqtt.subscribe, cron, drivers, etc.

**Comment by btsimonh** [comment]
thankyou @sfromis 
Can I call 
int32_t callBerryEventDispatcher(const char *type, const char *cmd, int32_t idx, const char *payload, uint32_t data_len) 
directly?

It looks like the berry event function will iterate through drivers, and find a matching method for 'event_type'

so 
`callBerryEventDispatcher("webcam", "frame", 0, "", 0);`
would call any function called "webcam" on any berry 'driver'.  If a driver returns non zero, it will not try further drivers, otherwise you can have it serviced by multiple drivers if you wish...

If I called this at 75hz, it feels a little heavy.  (I'll make a flag to turn this on...).
For processing motion data once per second, not too bad.

is this what you were thinking?

**Reply by sfromis** [reply]
Probably, but not having done it myself, I deliberately only posted a suggestive pointer, without "implementation details".
However, with much of Berry interfacing with Tasmota in a 50 ms cycle, I'd call it "experimental" if frame processing works well via Berry calls. Not saying that it won't, just that it may be pushing things a bit much.

**Reply by btsimonh** [reply]
Calling 
`callBerryEventDispatcher("webcam", "frame", 0, "", 0);`
and 
`callBerryEventDispatcher("webcam", "motion", 0, "43 15034", (len));`
works very well combined with:
```
class MyWebcamDriver
  #- create a method for adding a button to the main menu -#
  def webcam(cmd, idx, payload)
    print("webcam "+cmd+" payload:"+payload)
  end
end
```

no claims on framerate!  it's serviced in the FUNC_LOOP of the webcam driver.  I may consider putting the framecount in... so we know if we missed one.
